digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala

```diff
@@ -0,0 +1,609 @@
+                Some(Projection(oldStart, oldEnd, newStart, newEnd, vvhTimeStamp))
+              } else {
+                None
+              }
+
+      case _ => None
+    }
+  }
+
+  private def testAssetsContainSegment(lanes: Seq[PersistedLane], linkId: Long, mStart: Double, mEnd: Double,
+                                       vvhTimeStamp: Long): Boolean = {
+    val targetAssets = lanes.filter(a => a.linkId == linkId)
+    targetAssets.nonEmpty && !targetAssets.exists(a => a.vvhTimeStamp >= vvhTimeStamp) && targetAssets.exists(
+      a => GeometryUtils.covered((a.startMeasure, a.endMeasure),(mStart,mEnd)))
+  }
+
+  private def addSourceRoadLinkToChangeInfo(extensionChangeInfo: ChangeInfo, replacementChangeInfo: ChangeInfo) = {
+    def givenAndEqualDoubles(v1: Option[Double], v2: Option[Double]) = {
+      (v1, v2) match {
+        case (Some(d1), Some(d2)) => d1 == d2
+        case _ => false
+      }
+    }
+    def givenAndEqualLongs(v1: Option[Long], v2: Option[Long]) = {
+      (v1, v2) match {
+        case (Some(l1), Some(l2)) => l1 == l2
+        case _ => false
+      }
+    }
+    // Test if these change infos extend each other. Then take the small little piece just after tolerance value to test if it is true there
+    val (mStart, mEnd) = (givenAndEqualDoubles(replacementChangeInfo.newStartMeasure, extensionChangeInfo.newEndMeasure),
+      givenAndEqualDoubles(replacementChangeInfo.newEndMeasure, extensionChangeInfo.newStartMeasure)) match {
+      case (true, false) =>
+        (replacementChangeInfo.oldStartMeasure.get + laneFilter.AllowedTolerance,
+          replacementChangeInfo.oldStartMeasure.get + laneFilter.AllowedTolerance + laneFilter.MaxAllowedError)
+      case (false, true) =>
+        (Math.max(0.0, replacementChangeInfo.oldEndMeasure.get - laneFilter.AllowedTolerance - laneFilter.MaxAllowedError),
+          Math.max(0.0, replacementChangeInfo.oldEndMeasure.get - laneFilter.AllowedTolerance))
+      case (_, _) => (0.0, 0.0)
+    }
+
+    if (mStart != mEnd && extensionChangeInfo.vvhTimeStamp == replacementChangeInfo.vvhTimeStamp)
+      Option(extensionChangeInfo.copy(oldId = replacementChangeInfo.oldId, oldStartMeasure = Option(mStart), oldEndMeasure = Option(mEnd)))
+    else
+      None
+  }
+
+  def fetchExistingMainLanesByRoadLinks( roadLinks: Seq[RoadLink], removedLinkIds: Seq[Long]): Seq[PersistedLane] = {
+    val linkIds = roadLinks.map(_.linkId)
+    val existingAssets =
+      withDynTransaction {
+        dao.fetchMainLanesByLinkIds( linkIds ++ removedLinkIds)
+      }.filterNot(_.expired)
+    existingAssets
+  }
+
+  def fetchExistingLanesByRoadLinks(roadLinks: Seq[RoadLink], removedLinkIds: Seq[Long] = Seq()): Seq[PersistedLane] = {
+    val linkIds = roadLinks.map(_.linkId)
+    val existingAssets =
+      withDynTransaction {
+        dao.fetchLanesByLinkIds(linkIds ++ removedLinkIds)
+      }.filterNot(_.expired)
+    existingAssets
+  }
```

Comment:
You can pass the function as a parameter so it would be no need to have this two functions that are pratically the same

```scala
  def fetchExistingLanesByRoadLinks(roadLinks: Seq[RoadLink], removedLinkIds: Seq[Long] = Seq(), fetch: Long => Seq[PersistedLane]): Seq[PersistedLane] = {
      withDynTransaction {
        fetch(roadLinks.map(_.linkId) ++ removedLinkIds)
      }.filterNot(_.expired)
  }
```

Reply:
Don't see any vantange on that.
only more hard to read

Reply:
Apply partial suggestion